### PR TITLE
[Maybe types] Add MakeMaybe(...)

### PR DESCRIPTION
**Pull request by agnat**
Convert `Locals<T>` to `MaybeLocals<T>`. When called with a `MaybeLocal<T>` it is the identity function.

Inspired by #410.

### CODE REVIEW

**Comment by agnat** [comment]
Ready for review.

File: doc/maybe_types.md [hunk 1]
```diff
@@ -478,3 +479,22 @@ Signature:
 ```c++
 Nan::MaybeLocal<v8::Object> Nan::CloneElementAt(v8::Local<v8::Array> array, uint32_t index);
 ```
+
+<a name="api_nan_make_maybe"></a>
+### Nan::MakeMaybe()
+
+Converts a `v8::Local<>` into a `Nan::MaybeLocal<>`. When called with a `Nan::MaybeLocal<>` it just returns it's argument. This is useful in generic template code that builds on NAN.
```
Comment (kkoopa) [inline doc/maybe_types.md]: remove the apostrophe

File: nan.h [hunk 2]
```diff
@@ -2157,6 +2157,32 @@ struct Tap {
...
+
+template <typename T> struct Maybefier<MaybeLocal<T> > {
+  static MaybeLocal<T> convert(MaybeLocal<T> v) {
+    return v;
+  }
+};
+
+}  // end of namespace imp
+
+template <typename T, template <typename> class CouldbeMaybe>
```
Comment (kkoopa) [inline nan.h]: Is this valid in C++03? Any docs on it? Never seen the construct before and it is somewhat puzzling.

**Comment by kkoopa** [comment]
Looks good, apart from the documentation nit. However, wouldn't `EnsureMaybe` or such be a more fitting name? `MakeMaybe` is good, but is it the best possible name?

File: doc/maybe_types.md (hunk 1 above)
Comment (kkoopa) [inline doc/maybe_types.md]: Wraps a `v8::Local<>` in a `Nan::MaybeLocal<>` is better than 'converts'.

File: nan.h (hunk 2 above)
Comment (bnoordhuis) [inline nan.h]: It's a called a template template parameter: http://stackoverflow.com/questions/213761/what-are-some-uses-of-template-template-parameters-in-c

It's valid C++03 and I'm reasonably sure it's even valid C++98.

Comment (agnat) [inline nan.h]: What @bnoordhuis says.

IIRC, the STL container allocators rely on it pretty heavily...

**Comment by agnat** [comment]
The best possible name still is `maybefy(...)`... definitely. ;)

But I already had my fun with naming the implementation detail `Maybefier`. So I went with `MakeMaybe(...)` because it has precedent in both NAN and C++/STL... think `make_pair(...)` &c. However, I'm ok with `EnsureMaybe(...)` too. So, since I'm a `+0` it's up to you guys. Just tell me what it should be. ;)

**Comment by kkoopa** [comment]
`MakeMaybe` it is then. Go ahead and flatten and I'll merge once 2.1 gets near.

**Comment by agnat** [comment]
Actually, `EnsureMaybe(...)` is pretty funny too... lol

So, `MakeMaybe(...)` ... definitely.

**Comment by kkoopa** [comment]
Maybe `CouldbeMaybe` could be `MaybeMaybe`?

**Comment by agnat** [comment]
Exactly.

**Comment by kkoopa** [comment]
How much wood would a woodchuck chuck if a woodchuck would chuck wood?

File: nan.h (hunk 2 above)
Comment (agnat) [inline nan.h]: Ok, it's official. There is no good free text on template template parameters.

However, _The Text_ on the subject is [C++ Templates - The Complete Guide](http://www.josuttis.com/tmplbook/), a book every C++ programmer should have.

**Comment by agnat** [comment]
What's up with AppVeyor? It's just doing its windows thing, being unstable and all...? Or should I be worried?

**Comment by kkoopa** [comment]
Nothing to worry about. It is always flaky.

File: doc/maybe_types.md [hunk 3]
```diff
@@ -478,3 +479,22 @@ Signature:
...
+```c++
+  MaybeLocal<v8::Number> someNumber = MakeMaybe(New<v8::Number>(3.141592654));
+  MaybeLocal<v8::String> someString = MakeMaybe(New<v8::String>("probably"));
+```
+
+Signature:
+
+```c++
+template <typename T, template <typename> class CouldbeMaybe>
+Nan::MaybeLocal<T> Nan::MakeMaybe(CouldbeMaybe<T> v);
```
Comment (kkoopa) [inline doc/maybe_types.md]: This should say it takes a `MaybeMaybe<T>`.

**Comment by agnat** [comment]
Done.

**Comment by saper** [comment]
`CallMeMaybe` :)

By the way those AppVeyor failures are probably https://github.com/nodejs/nodejs.org/issues/167

**Comment by kkoopa** [comment]
Merged as 48d7b53d9702b0c7a060e69ea10fea8fb48d814d

**Pull request closed without merging**